Add LightGlue model

sbucaille opened the pull request:
# What does this PR do?

This PR implements [LightGlue](https://arxiv.org/pdf/2306.13643.pdf), a keypoint matching model that builds on top of [SuperPoint](https://huggingface.co/magic-leap-community/superpoint) and optimizes [SuperGlue](https://arxiv.org/pdf/1911.11763.pdf) inference time while setting SotA results in image matching tasks.
SuperGlue being currently implemented [through this PR](https://github.com/huggingface/transformers/pull/29886). This should be added after.
This branch is based on SuperGlue's branch and the PR is created before SuperGlue's is done to get CI tests results. Once SuperGlue's PR merged, this PR will be rebased onto main.

For now LightGlue is implemented like SuperGlue as a standalone model using `AutoModel` but this PR will also be the place where `AutoModelForKeypointMatching` will be created to group SuperGlue and LightGlue together.

This model supports FlashAttention2

Inference colab notebook: https://colab.research.google.com/drive/1c-6Hhe05vovKLctSFO7ZipDsHCzHSNuE?usp=sharing

## Who can review?

@amyeroberts  (cc @NielsRogge)

sbucaille commented:
@amyeroberts 
 
I had a weird bug with the `test_sdpa_can_compile_dynamic` test at first with this following message : 
```
C0710 21:51:39.831000 140316135185280 torch/_inductor/scheduler.py:781] [24/0] Error in codegen for ComputedBuffer(name='buf0', layout=AliasedLayout('cpu', torch.float16, size=[1], stride=[1]), data=Pointwise(
C0710 21:51:39.831000 140316135185280 torch/_inductor/scheduler.py:781] [24/0]   'cpu',
C0710 21:51:39.831000 140316135185280 torch/_inductor/scheduler.py:781] [24/0]   torch.float16,
C0710 21:51:39.831000 140316135185280 torch/_inductor/scheduler.py:781] [24/0]   def inner_fn(index):
C0710 21:51:39.831000 140316135185280 torch/_inductor/scheduler.py:781] [24/0]       _ = index
C0710 21:51:39.831000 140316135185280 torch/_inductor/scheduler.py:781] [24/0]       tmp0 = ops.index_expr(s2, torch.float16)
C0710 21:51:39.831000 140316135185280 torch/_inductor/scheduler.py:781] [24/0]       return tmp0
C0710 21:51:39.831000 140316135185280 torch/_inductor/scheduler.py:781] [24/0]   ,
C0710 21:51:39.831000 140316135185280 torch/_inductor/scheduler.py:781] [24/0]   ranges=[1],
C0710 21:51:39.831000 140316135185280 torch/_inductor/scheduler.py:781] [24/0]   origin_node=None,
C0710 21:51:39.831000 140316135185280 torch/_inductor/scheduler.py:781] [24/0]   origins={cat}
C0710 21:51:39.831000 140316135185280 torch/_inductor/scheduler.py:781] [24/0] ))

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

var = <torch._inductor.codegen.cpp.CppCSEVariable object at 0x7f63391083a0>
cache = {'c10::convert<half>(ks0)': <torch._inductor.codegen.cpp.CppCSEVariable object at 0x7f63391083a0>}

    def find_fp32_var(var, cache):
        fp32_cse_var = None
        fp32_cse_var_name = None
        lowp_dtype = None
        for expr, cse_var in cache.items():
            if cse_var == var:
                lowp_dtype = is_to_lowp_dtype(expr)
                if lowp_dtype:
                    m = re.search(r"tmp\d+", expr)
>                   assert m
E                   torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
E                   AssertionError: 
E                   
E                   
E                   You can suppress this exception and fall back to eager by setting:
E                       import torch._dynamo
E                       torch._dynamo.config.suppress_errors = True
```

This message pointed at an integer variable `height` that I am using to normalize keypoints. By adding the following lines before where it happens, I managed to pass the test, but is it a good way of doing it ? 
```python
height = torch.tensor(height, device=device)
width = torch.tensor(width, device=device)
```

For some reason, I can't manage to pass the following error with `test_sdpa_can_dispatch_on_flash` as it can't find flash attention backend although I am running the latest torch version 2.3.1, on WSL with a 4080 GPU.
```
    def forward(self, q: torch.Tensor, k: torch.Tensor, v: torch.Tensor) -> Tuple[torch.Tensor, None]:
>       attn_output = torch.nn.functional.scaled_dot_product_attention(q, k, v)
E       RuntimeError: No available kernel. Aborting execution.
```

By taking the valuable feedback on the SuperGlue PR, this LightGlue PR is now very close to a final state. Some docs are still needed here and there and I need to complete the model doc on https://huggingface.co/stevenbucaille/lightglue

sbucaille commented:
@amyeroberts 
 
I had a weird bug with the `test_sdpa_can_compile_dynamic` test at first with this following message : 
```
C0710 21:51:39.831000 140316135185280 torch/_inductor/scheduler.py:781] [24/0] Error in codegen for ComputedBuffer(name='buf0', layout=AliasedLayout('cpu', torch.float16, size=[1], stride=[1]), data=Pointwise(
C0710 21:51:39.831000 140316135185280 torch/_inductor/scheduler.py:781] [24/0]   'cpu',
C0710 21:51:39.831000 140316135185280 torch/_inductor/scheduler.py:781] [24/0]   torch.float16,
C0710 21:51:39.831000 140316135185280 torch/_inductor/scheduler.py:781] [24/0]   def inner_fn(index):
C0710 21:51:39.831000 140316135185280 torch/_inductor/scheduler.py:781] [24/0]       _ = index
C0710 21:51:39.831000 140316135185280 torch/_inductor/scheduler.py:781] [24/0]       tmp0 = ops.index_expr(s2, torch.float16)
C0710 21:51:39.831000 140316135185280 torch/_inductor/scheduler.py:781] [24/0]       return tmp0
C0710 21:51:39.831000 140316135185280 torch/_inductor/scheduler.py:781] [24/0]   ,
C0710 21:51:39.831000 140316135185280 torch/_inductor/scheduler.py:781] [24/0]   ranges=[1],
C0710 21:51:39.831000 140316135185280 torch/_inductor/scheduler.py:781] [24/0]   origin_node=None,
C0710 21:51:39.831000 140316135185280 torch/_inductor/scheduler.py:781] [24/0]   origins={cat}
C0710 21:51:39.831000 140316135185280 torch/_inductor/scheduler.py:781] [24/0] ))

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

var = <torch._inductor.codegen.cpp.CppCSEVariable object at 0x7f63391083a0>
cache = {'c10::convert<half>(ks0)': <torch._inductor.codegen.cpp.CppCSEVariable object at 0x7f63391083a0>}

    def find_fp32_var(var, cache):
        fp32_cse_var = None
        fp32_cse_var_name = None
        lowp_dtype = None
        for expr, cse_var in cache.items():
            if cse_var == var:
                lowp_dtype = is_to_lowp_dtype(expr)
                if lowp_dtype:
                    m = re.search(r"tmp\d+", expr)
>                   assert m
E                   torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
E                   AssertionError: 
E                   
E                   
E                   You can suppress this exception and fall back to eager by setting:
E                       import torch._dynamo
E                       torch._dynamo.config.suppress_errors = True
```

This message pointed at an integer variable `height` that I am using to normalize keypoints. By adding the following lines before where it happens, I managed to pass the test, but is it a good way of doing it ? 
```python
height = torch.tensor(height, device=device)
width = torch.tensor(width, device=device)
```

For some reason, I can't manage to pass the following error with `test_sdpa_can_dispatch_on_flash` as it can't find flash attention backend although I am running the latest torch version 2.3.1, on WSL with a 4080 GPU.
```
    def forward(self, q: torch.Tensor, k: torch.Tensor, v: torch.Tensor) -> Tuple[torch.Tensor, None]:
>       attn_output = torch.nn.functional.scaled_dot_product_attention(q, k, v)
E       RuntimeError: No available kernel. Aborting execution.
```

By taking the valuable feedback on the SuperGlue PR, this LightGlue PR is now very close to a final state. Some docs are still needed here and there and I need to complete the model doc on https://huggingface.co/stevenbucaille/lightglue

amyeroberts commented:
@sbucaille Great! 

For the sdpa test, converting ints to torch.tensors is definitely OK. There's a bunch of different functionalities which require this e.g. when using jit.trace

For the FA2 test, I'm not sure :/ It looks like an install issue, but I haven't seen before so uncertain

sbucaille commented:
@amyeroberts In the end I removed the sdpa support for now as I am running into a bug similar to [this issue](https://github.com/pytorch/pytorch/issues/133166), I guess I'll see later for that

Rocketknight1 commented:
@sbucaille feel free to let us know whenever this is ready for another review!

sbucaille commented:
Hi @Rocketknight1, thanks for the interest in my PR ! 
The code concerning LightGlue is close to its final form, I've tried to replicate all the feedback Amy and @qubvel provided me on the [SuperGlue's PR](https://github.com/huggingface/transformers/pull/29886), including the implementation of batching and using copies from existing transformers classes.
Although, I currently included in the branch the changes from SuperGlue's PR for two reasons : 
- For personal convenience, having SuperGlue's code while implementing LightGlue was helpful
- Since LightGlue is a keypoint matching like SuperGlue, I thought it would be worth to create an `AutoClassForKeypointMatching` including both models.

I still have minor things to fix such as this : 
![matched_image](https://github.com/user-attachments/assets/b8fc81f7-4ffc-4a38-81bf-2e67d3c7ae16)

I've also included a run-slow commit to check the integration tests, if one of you can trigger the action to see the result, but I suspect we will encounter the same problem as discussed [here](https://github.com/huggingface/transformers/pull/29886#issuecomment-2488479633)

I'd say I'll make another pass on the whole code, fix the several problems I mentionned and will ping you when it's ready ?

sbucaille commented:
@Rocketknight1 @qubvel PR is ready for a review, I still need to update `lightglue.md` and hub repo docs.
Should I contact the authors about whether they would like to have the model pushed to their own repo ? Part of LightGlue's authors are the same as SuperGlue's but LightGlue was not made by Magic Leap but a research team from Zurich. Also their licence is I believe fully open source as described [here](https://github.com/cvg/LightGlue?tab=Apache-2.0-1-ov-file#readme), I'll update the repo licence accordingly.

qubvel commented on src/transformers/models/lightglue/modeling_lightglue.py:
```diff
@@ -0,0 +1,1167 @@
+        self, keypoints: torch.Tensor, output_hidden_states: Optional[bool] = False
+    ) -> Union[Tuple[torch.Tensor], Tuple[torch.Tensor, torch.Tensor]]:
+        projected_keypoints = self.projector(keypoints)
+        cosines, sines = torch.cos(projected_keypoints), torch.sin(projected_keypoints)
```
```suggestion
        cosines = torch.cos(projected_keypoints)
        sines = torch.sin(projected_keypoints)
```

qubvel commented on src/transformers/models/lightglue/modeling_lightglue.py:
```diff
@@ -0,0 +1,1167 @@
+        self.final_projection = nn.Linear(self.descriptor_dim, self.descriptor_dim, bias=True)
+        self.matchability = nn.Linear(self.descriptor_dim, 1, bias=True)
+
+    def forward(self, descriptors: torch.Tensor, mask: torch.Tensor) -> torch.Tensor:
+        m_descriptors = self.final_projection(descriptors)
+        m_descriptors = m_descriptors / self.descriptor_dim**0.25
+        matchability = self.matchability(descriptors)
+        batch_size, num_keypoints, descriptor_dim = descriptors.shape
+        m_descriptors = m_descriptors.reshape(batch_size // 2, 2, num_keypoints, descriptor_dim)
+        m_descriptors0 = m_descriptors[:, 0]
+        m_descriptors1 = m_descriptors[:, 1]
+        similarity = torch.einsum("bmd,bnd->bmn", m_descriptors0, m_descriptors1)
+        if mask is not None:
+            mask = mask.reshape(batch_size // 2, 2, num_keypoints)
+            mask0 = mask[:, 0].unsqueeze(-1)
+            mask1 = mask[:, 1].unsqueeze(-1).transpose(-1, -2)
+            mask = mask0 * mask1
+            similarity = similarity.masked_fill(mask == 0, torch.finfo(similarity.dtype).min)
+
+        matchability = matchability.reshape(batch_size // 2, 2, num_keypoints, 1)
+        matchability_0 = matchability[:, 0]
+        matchability_1 = matchability[:, 1]
+        scores = sigmoid_log_double_softmax(similarity, matchability_0, matchability_1)
+        return scores
```
maybe add some comments

sbucaille replied:
Added comments to most forward methods

qubvel commented on src/transformers/models/lightglue/modeling_lightglue.py:
```diff
@@ -0,0 +1,1167 @@
+        return token
+
+
+def filter_matches(scores: torch.Tensor, threshold: float) -> Tuple[torch.Tensor, torch.Tensor]:
+    """obtain matches from a log assignment matrix [Bx M+1 x N+1]"""
+    batch_size, _, _ = scores.shape
+    max0, max1 = scores[:, :-1, :-1].max(2), scores[:, :-1, :-1].max(1)
+    matches0, matches1 = max0.indices, max1.indices
+    indices0 = torch.arange(matches0.shape[1], device=matches0.device)[None]
+    indices1 = torch.arange(matches1.shape[1], device=matches1.device)[None]
+    mutual0 = indices0 == matches1.gather(1, matches0)
+    mutual1 = indices1 == matches0.gather(1, matches1)
+    max0 = max0.values.exp()
+    zero = max0.new_tensor(0)
+    matching_scores0 = torch.where(mutual0, max0, zero)
+    matching_scores1 = torch.where(mutual1, matching_scores0.gather(1, matches1), zero)
+    valid0 = mutual0 & (matching_scores0 > threshold)
+    valid1 = mutual1 & valid0.gather(1, matches1)
+    matches0 = torch.where(valid0, matches0, -1)
+    matches1 = torch.where(valid1, matches1, -1)
+    matches = torch.stack([matches0, matches1]).transpose(0, 1).reshape(batch_size * 2, -1)
+    matching_scores = torch.stack([matching_scores0, matching_scores1]).transpose(0, 1).reshape(batch_size * 2, -1)
+
+    return matches, matching_scores
```
Also looks like lots code lines, would be nice to add some docs, comments

sbucaille replied:
Added comments to this function for more clarity

qubvel commented on src/transformers/models/lightglue/modeling_lightglue.py:
```diff
@@ -0,0 +1,1167 @@
+        super().__init__(config)
+
+        self.keypoint_detector = AutoModelForKeypointDetection.from_config(
+            config.keypoint_detector_config, attn_implementation="eager"
```
Is there any other attention implemented for SuperPoint? If no, we can remove `attn_implementation` arg here

sbucaille replied:
Indeed, removed the arg

qubvel commented on src/transformers/models/lightglue/modeling_lightglue.py:
```diff
@@ -0,0 +1,1167 @@
+
+        self.post_init()
+
+    def confidence_threshold(self, layer_index: int) -> float:
```
lets name it `get_confidence_threshold`

sbucaille replied:
Method renamed !

qubvel commented on src/transformers/models/lightglue/modeling_lightglue.py:
```diff
@@ -0,0 +1,1167 @@
+    Args:
+        loss (`torch.FloatTensor` of shape `(1,)`, *optional*):
+            Loss computed during training.
+        mask (`torch.BoolTensor` of shape `(batch_size, num_keypoints)`):
+            Mask indicating which values in matches and matching_scores are keypoint matching information.
+        matches (`torch.FloatTensor` of shape `(batch_size, 2, num_matches)`):
+            Index of keypoint matched in the other image.
+        matching_scores (`torch.FloatTensor` of shape `(batch_size, 2, num_matches)`):
+            Scores of predicted matches.
+        keypoints (`torch.FloatTensor` of shape `(batch_size, num_keypoints, 2)`):
+            Absolute (x, y) coordinates of predicted keypoints in a given image.
+        prune (`torch.IntTensor` of shape `(batch_size, num_keypoints)`):
```
I would recommend mask to be after `matching_scores`

sbucaille replied:
I reordered as the following : matches, scores, keypoints, prune then mask as mask is important for all the aforementioned arguments

qubvel commented on src/transformers/models/lightglue/__init__.py:
Thanks for using updated format!

qubvel commented on tests/models/lightglue/test_image_processing_lightglue.py:
```diff
@@ -0,0 +1,317 @@
+            image_inputs = [image_inputs[i : i + 2] for i in range(0, len(image_inputs), 2)]
+        return image_inputs
+
+    def prepare_keypoint_matching_output(self, pixel_values):
+        max_number_keypoints = 50
+        batch_size = len(pixel_values)
+        mask = torch.zeros((batch_size, 2, max_number_keypoints), dtype=torch.int)
```
Adding some comments on whats happening in this method would be nice

sbucaille replied:
I added some comments for this function

qubvel commented on tests/models/lightglue/test_modeling_lightglue.py:
If there are any tests copied from superglue please use `#Copied from ` as well

sbucaille replied:
Copies from Superglue were present indeed but mostly in the image_processing tests

qubvel reviewed (commented):
Hey @sbucaille! Thanks for working on this model. 

I made an initial review, not very in-depth, but overall it looks good to me. The main point is to make the code a bit easier to read by adding some comments and splitting it into stages. A lot of consecutive lines of code could be improved, and at least adding `\n` would be useful to indicate the end of a logical part. Additionally, comments/docstrings would also help.
And thank you for taking into account the comments from the superglue PR 🤗

sbucaille commented on src/transformers/models/lightglue/configuration_lightglue.py:
```diff
@@ -0,0 +1,102 @@
+        if descriptor_dim % num_heads != 0:
+            raise ValueError("descriptor_dim % num_heads is different from zero")
+
+        # TODO Add `add_scale_and_orientation` if keypoint detectors like SIFT or DogHardNet are implemented
```
This comment was made a long ago, but after looking at SIFT and DogHardNet, I don't think they'll be implemented. SIFT is a traditional technique that is implemented in OpenCV. DogHardNet is a deep technique but there is no benchmark in the original paper whether this is accurate or not.

qubvel replied:
Ok, lets remove TODO then

sbucaille replied:
I removed the TODO

qubvel commented on src/transformers/models/lightglue/image_processing_lightglue.py:
```diff
@@ -0,0 +1,588 @@
+        size: Dict[str, int] = None,
+        resample: PILImageResampling = PILImageResampling.BILINEAR,
+        do_rescale: bool = True,
+        rescale_factor: float = 1 / 255,
```
Ok, then it's required to add this parameter in order to control behavior

sbucaille replied:
I've changed the lightglue repo to be `lightglue_superpoint` with the configuration specific to SuperPoint usage, I've added do_grayscale as True in this config and set it False as default in the class

qubvel replied:
We usually align it with the config, if anyone will use it with DISK they can provide `do_grayscale=False` explicitly.

sbucaille replied:
Alright, I reverted it to make do_grayscale True by default in the image processor

sbucaille commented on src/transformers/models/lightglue/convert_lightglue_to_hf.py:
```diff
@@ -0,0 +1,250 @@
+    preprocessor = LightGlueImageProcessor()
+    inputs = preprocessor(images=images, return_tensors="pt").to("cuda")
+    model.to("cuda")
+    with torch.no_grad():
```
Fixed

sbucaille commented on src/transformers/models/lightglue/convert_lightglue_to_hf.py:
```diff
@@ -0,0 +1,250 @@
+    expected_matches_values = torch.tensor([-1, -1, -1, -1, -1, -1, -1, -1, 540, -1], dtype=torch.int64).to("cuda")
+    expected_matching_scores_values = torch.tensor([0, 0, 0.0167, 0.0304, 0.0328, 0, 0, 0.0095, 0.2964, 0.0352]).to(
+        "cuda"
+    )
```
Fixed aswell

sbucaille commented on src/transformers/models/lightglue/modeling_lightglue.py:
```diff
@@ -0,0 +1,1167 @@
+        attentions (`tuple(torch.FloatTensor)`, *optional*, returned when `output_attentions=True` is passed or when
+        `config.output_attentions=True`):
+            Tuple of `torch.FloatTensor` (one for each layer) of shape `(batch_size, 2, num_heads, num_keypoints,
+            num_keypoints)`.
```
Updated the docstring

sbucaille commented on src/transformers/models/lightglue/modeling_lightglue.py:
```diff
@@ -0,0 +1,1167 @@
+    matching_scores: torch.FloatTensor = None
+    keypoints: torch.FloatTensor = None
+    prune: torch.IntTensor = None
+    hidden_states: Optional[Tuple[torch.FloatTensor]] = None
```
Not sure what you mean with this comment here 👀

qubvel replied:
I mean if argument is `None` by default it should be annotated as Optional

```python
mask: Optional[torch.FloatTensor] = None
```

sbucaille replied:
Ah yes it makes sense, I'll replicate that to SuperGlue's PR as well

sbucaille commented:
Hi @qubvel,
I tried to answer most of your comments, I think pushing new commits made some of them "outdated" and then disappeared.
To recap the recent changes, I've addressed the comments, from small changes to bigger. 
I've tried to refactor most of the match_image_pair and forward methods, tried to use more list comprehension to repetitive instructions as well as used `nn.utils.rnn.pad_sequence` which I forgot the existence (and I think it could be useful for SuperGlue's PR as well) which allows me not necessarily instantiate a big tensor upfront and then filling it but rather batching together tensors with inconsistent shapes.
Still one thing missing that I will do tomorrow morning is the `lightglue.md`.
Other than that let me know what you think !

qubvel commented on src/transformers/models/lightglue/image_processing_lightglue.py:
```diff
@@ -0,0 +1,588 @@
+    return image
+
+
+def pad_images(images, pad_value=0, data_format: Optional[Union[str, ChannelDimension]] = None):
```
In case it's not intended to use a model with images of different shapes we can remove it. As far as I understand image processor will resize all images to the same size

sbucaille replied:
By default yes it will. I removed it

qubvel commented on src/transformers/models/lightglue/modeling_lightglue.py:
```diff
@@ -0,0 +1,1158 @@
+
+        descriptors = self_attention_output[0]
+
+        encoder_hidden_states = (
+            descriptors.reshape(-1, 2, num_keypoints, descriptor_dim)
+            .flip(1)
+            .reshape(batch_size, num_keypoints, descriptor_dim)
+        )
+        encoder_attention_mask = (
+            attention_mask.reshape(-1, 2, 1, 1, num_keypoints).flip(1).reshape(batch_size, 1, 1, num_keypoints)
+            if attention_mask is not None
+            else None
+        )
```
Would appretiate comments here on why we are doing flip/reshape

sbucaille replied:
Comments added !

qubvel commented on src/transformers/models/lightglue/modeling_lightglue.py:
```diff
@@ -0,0 +1,1158 @@
+    main_input_name = "pixel_values"
+    supports_gradient_checkpointing = False
+    _supports_flash_attn_2 = True
+    _supports_sdpa = False
```
There SDPA module, shkoud we turn it to `True`?

qubvel commented on src/transformers/models/lightglue/modeling_lightglue.py:
```diff
@@ -0,0 +1,1158 @@
+    @add_start_docstrings_to_model_forward(LIGHTGLUE_INPUTS_DOCSTRING)
+    def forward(
+        self,
+        pixel_values: torch.FloatTensor = None,
```
```suggestion
        pixel_values: torch.FloatTensor,
```

qubvel commented on src/transformers/models/lightglue/modeling_lightglue.py:
```diff
@@ -0,0 +1,1158 @@
+        )
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        if len(pixel_values.size()) != 5 or pixel_values.size(1) != 2:
```
```suggestion
        if pixel_values.ndim != 5 or pixel_values.size(1) != 2:
```

qubvel commented on src/transformers/models/lightglue/modeling_lightglue.py:
```diff
@@ -0,0 +1,1158 @@
+
+        batch_size, _, channels, height, width = pixel_values.shape
+
+        list_keypoint_detection = [self.keypoint_detector(image_pair) for image_pair in pixel_values]
+
+        max_keypoints = max([keypoint_detection[0].shape[1] for keypoint_detection in list_keypoint_detection])
+        keypoints = torch.zeros(
+            (batch_size, 2, max_keypoints, 2), device=pixel_values.device, dtype=pixel_values.dtype
+        )
+        descriptors = torch.zeros(
+            (batch_size, 2, max_keypoints, self.config.descriptor_dim),
+            device=pixel_values.device,
+            dtype=pixel_values.dtype,
+        )
+        mask = torch.zeros((batch_size, 2, max_keypoints), device=pixel_values.device, dtype=torch.int)
+
+        for i, keypoint_detection_output in enumerate(list_keypoint_detection):
+            _keypoints, _scores, _descriptors, _mask = keypoint_detection_output[:4]
+            keypoints[i, :, : _keypoints.shape[1], :] = _keypoints
+            descriptors[i, :, : _descriptors.shape[1], :] = _descriptors
+            mask[i, :, : _mask.shape[1]] = _mask
+
+        absolute_keypoints = keypoints.clone()
+        absolute_keypoints[:, :, :, 0] = absolute_keypoints[:, :, :, 0] * width
+        absolute_keypoints[:, :, :, 1] = absolute_keypoints[:, :, :, 1] * height
+
```
Can we do batch prediction as follows?

```python
pixel_values = pixel_values.reshape(-1, channels, height, width)
keypoints = self.keypoint_detector(pixel_values)
```
And then reshape keypoints/scores/etc

sbucaille replied:
I've changed the code to indeed use the keypoint_detector with batched input

qubvel commented on docs/source/en/model_doc/lightglue.md:
```diff
@@ -0,0 +1,148 @@
+plt.axis("off")
+
+# Retrieve the keypoints and matches
+output = outputs[0]
```
Getting error on this in line. Outputs is a LightGlue output object, not a list

sbucaille replied:
I can't seem to reproduce the error on my side

qubvel replied:
Works fine now

qubvel reviewed (commented):
Thanks for all the updates! I did another round of review, and overall, it looks great 🤗

sbucaille commented on docs/source/en/model_doc/lightglue.md:
````diff
@@ -0,0 +1,148 @@
+
+image0_matching_scores = outputs.matching_scores[0, 0][image0_indices]
+image1_matching_scores = outputs.matching_scores[0, 1][image1_indices]
+```
````
Makes sense, I updated the docs

sbucaille commented on src/transformers/models/lightglue/configuration_lightglue.py:
```diff
@@ -0,0 +1,108 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from ... import SuperPointConfig
```
👍

sbucaille commented on src/transformers/models/lightglue/configuration_lightglue.py:
```diff
@@ -0,0 +1,108 @@
+        filter_threshold (`float`, *optional*, defaults to 0.1):
+            The confidence threshold used to filter matches
+        initializer_range (`float`, *optional*, defaults to 0.02):
+            The standard deviation of the truncated_normal_initializer for initializing all weight matrices.
```
I updated the config to use these parameters and applied the changes on all the code, although should I also remove `descriptor_dim` and only use `hidden_size` when references are made in the model ?

qubvel replied:
I also thought about it, but `descriptor_dim` sounds clearer to me. Feel free to use whichever one you think is better.

sbucaille replied:
That's also what I had in mind, `descriptor_dim` makes sense from `match_image_pair` standpoint, it makes the code more understandable

sbucaille commented on src/transformers/models/lightglue/convert_lightglue_to_hf.py:
```diff
@@ -0,0 +1,254 @@
+    state_dict = add_keypoint_detector_state_dict(state_dict)
+
+    print("Loading the checkpoint in a SuperGlue model...")
+    with torch.device("cuda"):
```
Indeed forgot this instance of cuda, I added it as a paramter of verify_model_outputs, to use the same variable along the script execution

sbucaille commented on src/transformers/models/lightglue/convert_lightglue_to_hf.py:
```diff
@@ -0,0 +1,254 @@
+    verify_model_outputs(model)
+    print("Model outputs verified successfully.")
+
+    organization = "stevenbucaille"
```
`organization` added as script arg

sbucaille commented on src/transformers/models/lightglue/modeling_lightglue.py:
```diff
@@ -0,0 +1,1158 @@
+
+
+class LightGlueMLP(nn.Module):
+    def __init__(self, config):
```
I've added missing `LightGlueConfig` typehints to __init__'s

qubvel replied:
Thanks, a few more missed: there is not type hint in this method and for `LightGlueSelfOutput.__init__` :)

sbucaille commented on src/transformers/models/lightglue/modeling_lightglue.py:
```diff
@@ -0,0 +1,1158 @@
+        embeddings_dim = config.descriptor_dim
+        self.dense = nn.Linear(2 * embeddings_dim, 2 * embeddings_dim)
+        self.layer_norm = nn.LayerNorm(2 * embeddings_dim, elementwise_affine=True)
+        self.activation = ACT2FN["gelu"]
```
👍

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_31718). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

sbucaille commented:
@qubvel thanks for triggering the slow tests. Is there any guidelines for installing transformers with flash-attn such that I don't get the following error on `test_sdpa_can_dispatch_on_flash` : 
```
>       attn_output = torch.nn.functional.scaled_dot_product_attention(
            query_layer,
            key_layer,
            value_layer,
            attn_mask=attention_mask,
            dropout_p=self.dropout_prob if self.training else 0.0,
        )
E       RuntimeError: No available kernel. Aborting execution.
```
On a second setup I use I have the same error, the first was on WSL, the second is Ubuntu 🤔 

Anyway, looks like there is another bug triggered by the grid_sample instruction in SuperPoint, after a small search I found this [issue](https://github.com/pytorch/pytorch/issues/127224) that may apply to us but I can't replicate this error on my setup even by downgrading torch version

qubvel commented on src/transformers/models/lightglue/configuration_lightglue.py:
```diff
@@ -0,0 +1,115 @@
+        self.keypoint_detector_config = keypoint_detector_config
+
+        self.hidden_size = descriptor_dim
+        self.attention_probs_dropout_prob = 0
+        self.rotary_value = False
+        self.hidden_act = "gelu"
```
Let's add this to init function args well

qubvel commented on src/transformers/models/lightglue/modeling_lightglue.py:
```diff
@@ -0,0 +1,1155 @@
+                    # If a pair of images is considered early stopped, we compute the matches for the remaining
+                    # keypoints and stop the forward pass through the transformer layers for this pair of images.
+                    early_stops = early_stopped_pairs.repeat_interleave(2)
+                    _, layer_num_keypoints = mask.shape
```
It seems not used anywhere

qubvel reviewed (commented):
Hi @sbucaille! Thanks for addressing the comments. I took a quick look and everything looks great to me! I've also added the inference notebook to the PR's initial message. I will conduct a more in-depth review later this week, and if everything is fine, I will pass it to the next stage. 🤗

qubvel commented:
Re flash attention, do you have https://github.com/Dao-AILab/flash-attention installed? Idk if thats required, also it might be that a specific version of GPU is required (A or H architecture).

sbucaille commented:
Hi @qubvel , hope you are doing well ! 
Let me gently bump this PR as I have several remarks regarding the PR :
- The failing test should not be a problem once pytorch release a new version I believe as their [fix](https://github.com/pytorch/pytorch/pull/134812) has been merged.
- Just pushed a commit that makes verifying model outputs only if we provide the official LightGlue checkpoint url. With the release of the [MINIMA](https://github.com/LSXI7/MINIMA) version of LightGlue, we wouldn't be able to convert this version of LightGlue because the verification would naturally fails as the weights are different. So I've set a default variable with the official weights

About this MINIMA version, I guess I will reach out with them when the PR is merged so they can upload their own version on the hub so that it can be used with transformers. lmkwyt !

sbucaille commented:
Hi @qubvel,
Rebased the branch on main as SuperGlue's got merged.
I've applied the latest suggestions aswell.
I'm not sure what triggered the review request to this many people though, I just force pushed😅

qubvel commented:
No worries, @sbucaille! This is a CODEOWNERS experiment in `transformers`. We were trying to automate with rewiev ping with GitHub tools, but there seemed to be too many false positives, so it's going to be canceled. 

Will review PR this week, as far as I remember it's in a good shape, and thanks for updating with recent comments 🤗

ArthurZucker commented on docs/source/en/model_doc/lightglue.md:
```diff
@@ -0,0 +1,137 @@
+keypoints1_x, keypoints1_y = keypoints1[:, 0].numpy(), keypoints1[:, 1].numpy()
+
+# Plot the matches
+for keypoint0_x, keypoint0_y, keypoint1_x, keypoint1_y, matching_score in zip(
+        keypoints0_x, keypoints0_y, keypoints1_x, keypoints1_y, matching_scores
+):
+    plt.plot(
+        [keypoint0_x, keypoint1_x + image1.width],
+        [keypoint0_y, keypoint1_y],
+        color=plt.get_cmap("RdYlGn")(matching_score.item()),
+        alpha=0.9,
+        linewidth=0.5,
+    )
+    plt.scatter(keypoint0_x, keypoint0_y, c="black", s=2)
+    plt.scatter(keypoint1_x + image1.width, keypoint1_y, c="black", s=2)
```
as I mentioned last time, we can have a soft dependency and make sure that the processor allows users to interact with the images, both on the pre-processing and post-processing aspect

ArthurZucker commented on src/transformers/models/lightglue/__init__.py:
```diff
@@ -0,0 +1,28 @@
+# Copyright 2024 The HuggingFace Team. All rights reserved.
```
```suggestion
# Copyright 2025 The HuggingFace Team. All rights reserved.
```

sbucaille replied:
Ah yes, copyrights are updated throughout the changed files

ArthurZucker commented on src/transformers/models/lightglue/modeling_lightglue.py:
```diff
@@ -0,0 +1,1149 @@
+_CHECKPOINT_FOR_DOC_ = "stevenbucaille/lightglue_superpoint"
+
+
+def normalize_keypoints(keypoints: torch.Tensor, height: int, width: int) -> torch.Tensor:
```
this looks exactly the same as superglue

sbucaille replied:
Unfortunately it is not, I already checked in the past whether it would work to have the same but the result of normalization is different.
Above is SuperGlue, below is LightGlue
![image](https://github.com/user-attachments/assets/2908de7b-2775-4eb1-917a-e35ca8397d59)

ArthurZucker commented on src/transformers/models/lightglue/modeling_lightglue.py:
```diff
@@ -0,0 +1,1149 @@
+
+
+@dataclass
+class LightGlueKeypointMatchingOutput(ModelOutput):
```
same here, missing copied from. Or use modular!

sbucaille replied:
Compared to SuperGlue's output, there is an additionnal "prune" tensor

ArthurZucker commented on src/transformers/models/lightglue/modeling_lightglue.py:
```diff
@@ -0,0 +1,1149 @@
+        return output
+
+
+# Copied from transformers.models.roformer.modeling_roformer.RoFormerSelfAttention with RoFormer->LightGlue
```
I don't think we want to base it of Roformer, this depends on how involve it is for you but it's one of the "ugliest" code we have.

ArthurZucker commented on src/transformers/models/lightglue/modeling_lightglue.py:
```diff
@@ -0,0 +1,1149 @@
+        self.dropout = nn.Dropout(config.attention_probs_dropout_prob)
+
+        self.is_decoder = config.is_decoder
+        self.rotary_value = config.rotary_value
```
is rotary used or not? is this None or no?

sbucaille replied:
This attribute was a consequence of having a copy of Roformer

ArthurZucker commented on src/transformers/models/lightglue/modeling_lightglue.py:
```diff
@@ -0,0 +1,1149 @@
+        return outputs
+
+    @staticmethod
+    def apply_rotary_position_embeddings(sinusoidal_pos, query_layer, key_layer, value_layer=None):
```
newer models have a much simpler way to do this!

ArthurZucker commented on src/transformers/models/lightglue/modeling_lightglue.py:
```diff
@@ -0,0 +1,1149 @@
+        bsz, tgt_len, _ = hidden_states.size()
+
+        query_layer = self.transpose_for_scores(self.query(hidden_states))
+
+        is_cross_attention = encoder_hidden_states is not None
+
+        current_states = encoder_hidden_states if is_cross_attention else hidden_states
+        attention_mask = encoder_attention_mask if is_cross_attention else attention_mask
+
+        key_layer = self.transpose_for_scores(self.key(current_states))
+        value_layer = self.transpose_for_scores(self.value(current_states))
+        if sinusoidal_pos is not None:
+            if self.rotary_value:
+                query_layer, key_layer, value_layer = self.apply_rotary_position_embeddings(
+                    sinusoidal_pos, query_layer, key_layer, value_layer
+                )
+            else:
+                query_layer, key_layer = self.apply_rotary_position_embeddings(sinusoidal_pos, query_layer, key_layer)
+
+        # SDPA with memory-efficient backend is broken in torch==2.1.2 when using non-contiguous inputs and a custom
+        # attn_mask, so we need to call `.contiguous()` here. This was fixed in torch==2.2.0.
+        # Reference: https://github.com/pytorch/pytorch/issues/112577
+        if self.require_contiguous_qkv and query_layer.device.type == "cuda" and attention_mask is not None:
+            query_layer = query_layer.contiguous()
+            key_layer = key_layer.contiguous()
+            value_layer = value_layer.contiguous()
+
+        attn_output = torch.nn.functional.scaled_dot_product_attention(
+            query_layer,
+            key_layer,
+            value_layer,
+            attn_mask=attention_mask,
+            dropout_p=self.dropout_prob if self.training else 0.0,
+        )
+
+        attn_output = attn_output.transpose(1, 2)
+        attn_output = attn_output.reshape(bsz, tgt_len, self.all_head_size)
+
+        outputs = (attn_output,)
+        return outputs
```
we no longer define classes but rather define different attention functions! See `modeling_llama.py`!

ArthurZucker commented on src/transformers/models/lightglue/modeling_lightglue.py:
```diff
@@ -0,0 +1,1149 @@
+    def __init__(self, config: LightGlueConfig):
+        super().__init__()
+        embeddings_dim = config.descriptor_dim
+        self.dense = nn.Linear(2 * embeddings_dim, 2 * embeddings_dim)
+        self.layer_norm = nn.LayerNorm(2 * embeddings_dim, elementwise_affine=True)
+        self.activation = ACT2FN[config.hidden_act]
+        self.output = nn.Linear(2 * embeddings_dim, embeddings_dim)
```
this looks classique and we usually define intermediate dim for the dimension of the intermediate MLP.

ArthurZucker commented on src/transformers/models/lightglue/modeling_lightglue.py:
```diff
@@ -0,0 +1,1149 @@
+        self.depth_confidence = config.depth_confidence
+        self.width_confidence = config.width_confidence
+
+        if self.descriptor_dim != config.keypoint_detector_config.descriptor_decoder_dim:
+            self.input_projection = nn.Linear(
+                config.keypoint_detector_config.descriptor_decoder_dim, self.descriptor_dim, bias=True
+            )
+        else:
+            self.input_projection = nn.Identity()
```
we want to avoid codepaths like this. what does the released checkpoint use?

sbucaille replied:
When SuperPoint is used as a descriptor, the dimension is 256, whether when it is the DISK descriptor (similar model to SuperPoint), it is 128 iirc as LightGlue got released with both descriptors, they needed this additional layer. Should I remove this linear projection for the moment and maybe when DISK is added in transformers in the future we had it back ?

ArthurZucker reviewed (commented):
Nice addition ! Let's make sure the API is standardized for keypoint matching models 🔥

sbucaille commented:
Thanks @ArthurZucker for your review, I realize this contribution falls short to your standards regarding modular and the different copies.
Let me work on that again, I propose to come up with a modular version based on more recent models like llama.
I'll let you know when it is ready for a new review !

sbucaille commented:
@ArthurZucker @qubvel I freshened up the PR with the use of modular (super easy to use btw nice work on that !).
Mainly used llama as suggested, although the forward is a bit modified to perform cross attention.
Let me know what you think of this first try !

sbucaille commented:
Added plotting in a new method in the image processor with matplotlib soft dependency in the utils files

sbucaille commented:
@ArthurZucker just a gentle bump here 🤗

ArthurZucker commented:
Hahah sorry @sbucaille catching up on my late notifications!

sbucaille commented:
Hi @ArthurZucker, no worries ! Since the authors created their organization on the Hub (https://huggingface.co/ETH-CVG), I updated the convert script to upload the model on a more official repo.
I also fixed the small merge conflicts and rebased on main

sbucaille commented:
Hi @ArthurZucker, just rebased the branch on main after merge conflicts appeared. Also added a small line with `require_backends` to the post processing method as I saw other post processing methods (such as RT DETR image processor) included it.

sbucaille commented on docs/source/en/model_doc/lightglue.md:
```diff
@@ -0,0 +1,104 @@
+
+-->
+
+# LightGlue
```
Got it, I'll keep it in mind for future PR's thanks !

ArthurZucker commented on src/transformers/models/lightglue/modular_lightglue.py:
```diff
@@ -0,0 +1,1000 @@
+from dataclasses import dataclass
```
missing licence

ArthurZucker commented on src/transformers/models/lightglue/modular_lightglue.py:
```diff
@@ -0,0 +1,1000 @@
+    return torch.stack((-x2, x1), dim=-1).flatten(start_dim=-2)
+
+
+class LightGlueAttention(LlamaAttention):
```
nice

ArthurZucker reviewed (approved):
Sorry for the awaited review, LGTM! 🤗

sbucaille committed:
init

sbucaille committed:
chore: various changes to LightGlue

sbucaille committed:
chore: various changes to LightGlue

sbucaille committed:
chore: various changes to LightGlue

sbucaille committed:
chore: various changes to LightGlue

sbucaille committed:
fix: add matches different from -1 to compute valid matches in post_process_keypoint_matching

sbucaille committed:
tests: added CUDA proof tests similar to SuperGlue

Commit:
chore: Applied some suggestions from review

- Added missing typehints
- Refactor "cuda" to device variable
- Variable renaming
- LightGlue output order changed
- Make style

sbucaille committed:
chore: various changes to modeling_lightglue.py

- Added "Copies from" statements for copied functions from modeling_superglue.py
- Added missing docstrings
- Removed unused functions or classes
- Removed unnecessary statements
- Added missing typehints
- Added comments to the main forward method

sbucaille committed:
[run-slow] lightglue

sbucaille committed:
Apply suggestions from code review

Co-authored-by: Pavel Iakubovskii <qubvel@gmail.com>

sbucaille committed:
chore: make style

sbucaille committed:
chore: simplified convert_lightglue_to_hf conversion map

sbucaille committed:
feat: adding batching implementation

sbucaille committed:
chore: make style

sbucaille committed:
chore: changed convert_lightglue_to_hf script to comply with new standard

sbucaille committed:
fix: changed lightglue HF repo to lightglue_superpoint with grayscale default to True

Commit:
feat: added threshold to post_process_keypoint_matching method

sbucaille committed:
[run-slow] lightglue

sbucaille committed:
chore: fixed imports in lightglue files

Commit:
[run-slow] lightglue

Commit:
chore: make style

Commit:
fix: added typehint and docs

Commit:
fix: added missing instructions that turns keypoints back to absolute coordinate before matching forward

sbucaille committed:
chore: various changes to convert_lightglue_to_hf.py

- Added model saving
- Added model reloading

Commit:
[run-slow] lightglue

sbucaille committed:
chore: changed lightglue test values to match new lightglue config pushed to hub

Commit:
fix: added missing grayscale argument in image processor in case use of SuperPoint keypoint detector

sbucaille committed:
feat: added validation for preprocess and post process method to LightGlueImageProcessor

Commit:
refactor: removed unused line

sbucaille committed:
refactor: make keypoints `(batch_size, num_keypoints, keypoint_dim)` through forward and unsqueeze only before attention layer

sbucaille committed:
[run-slow] lightglue

sbucaille committed:
chore: added "gelu" in LightGlueConfig as hidden_act parameter

sbucaille committed:
chore: added some docs and refactoring

sbucaille committed:
docs: removed unnecessary code in docs

sbucaille committed:
refactor: reordered _concat_early_stopped_outputs in LightGlue class

Commit:
refactor: applied suggestions from SuperGlue PR

sbucaille committed:
docs: added lightglue.md model doc

sbucaille committed:
docs: import SuperPointConfig only from a TYPE_CHECKING context

sbucaille committed:
chore: added organization as arg in convert_lightglue_to_hf.py script

sbucaille committed:
Fixed dynamo bug and image padding tests

sbucaille committed:
chore: use PretrainedConfig arguments `num_hidden_layers` and `num_attention_heads` instead of `num_layers` and `num_heads`

Commit:
refactor: modular LightGlue

Commit:
docs: added missing LightGlueConfig typehint in init methods

sbucaille committed:
Apply suggestions from code review

Co-authored-by: Pavel Iakubovskii <qubvel@gmail.com>

Commit:
refactor: added checkpoint url as default variable to verify models output only if it is the default url

sbucaille committed:
Apply suggestions from code review

Co-authored-by: Pavel Iakubovskii <qubvel@gmail.com>

sbucaille committed:
refactor: various refactoring to modeling_lightglue.py

- Removed unused functions
- Renamed variables for consistency
- Added comments for clarity
- Set methods to private in LightGlueForKeypointMatching
- Replaced tensor initialization to list then concatenation
- Used more pythonic list comprehension for repetitive instructions

sbucaille committed:
[run-slow] lightglue

sbucaille committed:
refactor: added comments and renamed filter_matches to get_matches_from_scores

sbucaille committed:
docs: added Optional typehint to LightGlueKeypointMatchingOutput

sbucaille committed:
fix: added fix for SDPA tests

sbucaille committed:
chore: set do_grayscale default value to True in LightGlueImageProcessor

sbucaille committed:
fix: removed unnecessary import

sbucaille committed:
refactor: used batched inference for keypoint detector forward pass

sbucaille committed:
docs: changed copyright to 2025

sbucaille committed:
chore: removed pad_images function

sbucaille committed:
docs: added missing LightGlueConfig typehint in nn.Module __init__ methods

sbucaille committed:
tests: removed sdpa support and changed expected values

sbucaille committed:
refactor: applied refactoring changes from SuperGlue's concat, batch and stack functions to LightGlue file

sbucaille committed:
tests: added tests with no early stop and keypoint pruning

sbucaille committed:
docs: fixed docstring for LightGlueImageProcessor

Commit:
fix: got rid of confidence_thresholds as registered buffers

sbucaille committed:
tests: added copied from statement with superglue tests

Commit:
fix: moved print message inside if statement

sbucaille committed:
docs: added comment to prepare_keypoint_matching_output function in tests

Commit:
refactor: added missing arguments in LightGlueConfig init method

sbucaille committed:
[run-slow] lightglue

sbucaille committed:
docs: added comments to reshape.flip.reshape instruction to perform cross attention

sbucaille committed:
refactor: refactor do_layer_keypoint_pruning

sbucaille committed:
refactor: set device variable

Commit:
fix: added log assignment r removal in convert script

sbucaille committed:
feat: added plot_keypoint_matching method to LightGlueImageProcessor with matplotlib soft dependency

Commit:
Updated convert script to push on ETH org

sbucaille committed:
fix: added missing import error for matplotlib

sbucaille committed:
chore: fixed copy to superpoint.image_processing_superpoint.convert_to_grayscale

sbucaille committed:
feat: adding batch implementation

sbucaille committed:
fix: added missing licence

sbucaille committed:
fix: make fix-copies

sbucaille commented:
Hi @ArthurZucker thanks for the review ! @Cyrilvallez minor changes are made and branch is rebased on main

sbucaille commented:
@Cyrilvallez rebased again due to another merge conflict

Cyrilvallez commented on src/transformers/models/lightglue/modular_lightglue.py:
```diff
@@ -0,0 +1,1013 @@
+        return output
+
+
+def rotate_half(x: torch.Tensor) -> torch.Tensor:
+    x1 = x[..., ::2]
+    x2 = x[..., 1::2]
+    return torch.stack((-x2, x1), dim=-1).flatten(start_dim=-2)
```
This can be imported directly from Cohere!

sbucaille replied:
Indeed it is the same implementation, I changed to use cohere implementation in modular

Cyrilvallez commented on src/transformers/models/lightglue/modular_lightglue.py:
```diff
@@ -0,0 +1,1013 @@
+    def __init__(self, config: LightGlueConfig):
+        super().__init__()
+        hidden_size = config.hidden_size
+        self.dense = nn.Linear(2 * hidden_size, 2 * hidden_size)
+        self.layer_norm = nn.LayerNorm(2 * hidden_size, elementwise_affine=True)
+        self.activation = ACT2FN[config.hidden_act]
+        self.output = nn.Linear(2 * hidden_size, hidden_size)
```
Let's use `intermediate_size` in the config instead of hard-coding the factor 2 here! Then, the whole class can be directly inherited from another model if we rename the layers accordingly!

sbucaille replied:
Config changed to add `intermediate_size`, although I couldn't find exactly the MLP implementation that could replace the whole class, but I used CLIPMLP as a base now

Cyrilvallez commented on src/transformers/models/lightglue/modular_lightglue.py:
```diff
@@ -0,0 +1,1013 @@
+        output_attentions: bool = False,
+    ) -> Tuple[torch.Tensor, Optional[Tuple[torch.Tensor]], Optional[torch.Tensor]]:
+        all_hidden_states = () if output_hidden_states else None
+        attention_outputs = self.attention(
```
Let's explicitly return both arguments here instead of slicing later

sbucaille replied:
Changed slicing to explicit variable names

Cyrilvallez commented on src/transformers/models/lightglue/modular_lightglue.py:
```diff
@@ -0,0 +1,1013 @@
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+        )
+        if not return_dict:
+            return tuple(
+                v
+                for v in [matches, matching_scores, keypoints, prune, mask, hidden_states, attentions]
+                if v is not None
+            )
```
to remove and use decorator `@can_return_tuple` instead!

Cyrilvallez commented on src/transformers/models/lightglue/modular_lightglue.py:
```diff
@@ -0,0 +1,1013 @@
+        labels: Optional[torch.LongTensor] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
```
to remove as well, see previous comment 🤗

sbucaille replied:
👍

Cyrilvallez commented on src/transformers/models/lightglue/modular_lightglue.py:
```diff
@@ -0,0 +1,1013 @@
+            all_hidden_states = all_hidden_states + (descriptors,)
+
+        batch_size, num_keypoints, descriptor_dim = descriptors.shape
+        self_attention_output = self.self_attention_block(
```
let's be explicit about all args here instead of slicing later!

sbucaille replied:
Made variables explicit here aswell

Cyrilvallez reviewed (commented):
Hey! Nice work! Are are some small details that we can still improve 🤗

sbucaille committed:
refactor: use cohere apply_rotary_pos_emb function

sbucaille committed:
fix: update model references to use ETH-CVG/lightglue_superpoint

sbucaille committed:
refactor: add and use intermediate_size attribute in config to inherit CLIPMLP for LightGlueMLP

sbucaille committed:
refactor: explicit variables instead of slicing

sbucaille committed:
refactor: use can_return_tuple decorator in LightGlue model

sbucaille committed:
fix: make fix-copies

sbucaille commented:
Hi @Cyrilvallez , I addressed your comments, let me know what you think !

Cyrilvallez commented on src/transformers/models/lightglue/modular_lightglue.py:
```diff
@@ -0,0 +1,995 @@
+
+        attention_interface: Callable = eager_attention_forward
+        if self.config._attn_implementation != "eager":
+            if self.config._attn_implementation == "sdpa" and kwargs.get("output_attentions", False):
+                logger.warning_once(
+                    "`torch.nn.functional.scaled_dot_product_attention` does not support `output_attentions=True`. Falling back to "
+                    'eager attention. This warning can be removed using the argument `attn_implementation="eager"` when loading the model.'
+                )
+            else:
+                attention_interface = ALL_ATTENTION_FUNCTIONS[self.config._attn_implementation]
```
We removed this check

```suggestion
           attention_interface = ALL_ATTENTION_FUNCTIONS[self.config._attn_implementation]
```

Cyrilvallez reviewed (commented):
Hey! Super super sorry for the long due review! This PR got buried somehow... 🥲 Don't hesitate to ping us once in a while if we're not being responsive enough! In the meantime, here is a new round to try to polish as much as possible! 🤗

sbucaille committed:
docs: Update model references in `lightglue.md` to use the correct pretrained model from ETH-CVG

sbucaille committed:
Refactor LightGlue configuration and processing classes

- Updated type hints for `keypoint_detector_config` in `LightGlueConfig` to use `SuperPointConfig` directly.
- Changed `size` parameter in `LightGlueImageProcessor` to be optional.
- Modified `position_embeddings` in `LightGlueAttention` and `LightGlueAttentionBlock` to be optional tuples.
- Cleaned up import statements across multiple files for better readability and consistency.

sbucaille committed:
refactor: Update LightGlue configuration to enforce eager attention implementation

- Added `attn_implementation="eager"` to `keypoint_detector_config` in `LightGlueConfig` and `LightGlueAttention` classes.
- Removed unnecessary logging related to attention implementation fallback.
- Cleaned up import statements for better readability.

sbucaille committed:
refactor: renamed message into attention_output

sbucaille committed:
fix: ensure device compatibility in LightGlueMatchAssignmentLayer descriptor normalization

- Updated the normalization of `m_descriptors` to use the correct device for the tensor, ensuring compatibility across different hardware setups.

sbucaille committed:
refactor: removed Conv layers from init_weights since LightGlue doesn't have any

sbucaille committed:
refactor: replace add_start_docstrings with auto_docstring in LightGlue models

- Updated LightGlue model classes to utilize the new auto_docstring utility for automatic documentation generation.
- Removed legacy docstring handling to streamline the code and improve maintainability.

sbucaille committed:
refactor: simplify LightGlue image processing tests by inheriting from SuperGlue

- Refactored `LightGlueImageProcessingTester` and `LightGlueImageProcessingTest` to inherit from their SuperGlue counterparts, reducing code duplication.
- Removed redundant methods and properties, streamlining the test setup and improving maintainability.

sbucaille commented on docs/source/en/model_doc/lightglue.md:
```diff
@@ -0,0 +1,104 @@
+images = [image1, image2]
+
+processor = AutoImageProcessor.from_pretrained("stevenbucaille/lightglue_superpoint")
+model = AutoModel.from_pretrained("stevenbucaille/lightglue_superpoint")
```
Changed it for the [official ones](https://huggingface.co/ETH-CVG/lightglue_superpoint)

sbucaille commented on src/transformers/models/lightglue/modular_lightglue.py:
````diff
@@ -0,0 +1,995 @@
+        >>> configuration = model.config
+        ```
+    """
+
+    model_type = "lightglue"
````
Added the attribute

sbucaille commented on src/transformers/models/lightglue/modular_lightglue.py:
```diff
@@ -0,0 +1,995 @@
+        output_attentions: bool = False,
+    ) -> Tuple[torch.Tensor, Optional[Tuple[torch.Tensor]], Optional[torch.Tensor]]:
+        all_hidden_states = () if output_hidden_states else None
+        message, attention_probs = self.attention(
```
Name changed

sbucaille commented on src/transformers/models/lightglue/modular_lightglue.py:
```diff
@@ -0,0 +1,995 @@
+
+    def _init_weights(self, module: nn.Module) -> None:
+        """Initialize the weights"""
+        if isinstance(module, (nn.Linear, nn.Conv2d, nn.Conv1d)):
```
Copy paste leftovers 🙈 Removed it !

sbucaille commented on tests/models/lightglue/test_image_processing_lightglue.py:
```diff
@@ -0,0 +1,393 @@
+
+
+class LightGlueImageProcessingTester(unittest.TestCase):
+    # Copied from tests.models.superglue.test_image_processing_superglue.SuperGlueImageProcessingTester.__init__
```
Same here

sbucaille committed:
test: forced eager attention implementation to LightGlue model tests

- Updated `LightGlueModelTester` to include `attn_implementation="eager"` in the model configuration.
- This change aligns the test setup with the recent updates in LightGlue configuration for eager attention.

sbucaille commented:
Hi @Cyrilvallez ,
Addressed most of your comments, there is still one concerning the ``attn_implementation`` where I'd like your input.
Although, I have another problem regarding using modular with SuperGlueImageProcessor. After tweaking my code for several hours, I can't figure out how to do it. Let me explain.
If I follow strictly like I did in ``modular_lightglue.py``, I naively write this, similarly to what I do with ``LlamaAttention`` : 
```py
from ..llama.modeling_llama import LlamaAttention, eager_attention_forward # To show that this import path works 
from ..superglue.image_processing_superglue import SuperGlueImageProcessor
class LightGlueImageProcessor(SuperGlueImageProcessor): ...
```
But running ``python utils/modular_model_converter.py --files_to_parse lightglue`` leads to this error : 
```
Converting src/transformers/models/lightglue/modular_lightglue.py to a single model single file format
Traceback (most recent call last):
  File "/home/steven/transformers/utils/modular_model_converter.py", line 1810, in <module>
    converted_files = convert_modular_file(file_name)
  File "/home/steven/transformers/utils/modular_model_converter.py", line 1736, in convert_modular_file
    wrapper.visit(cst_transformers)
  File "/home/steven/transformers/.venv/lib/python3.10/site-packages/libcst/metadata/wrapper.py", line 204, in visit
    return self.module.visit(visitor)
  File "/home/steven/transformers/.venv/lib/python3.10/site-packages/libcst/_nodes/module.py", line 89, in visit
    result = super(Module, self).visit(visitor)
  File "/home/steven/transformers/.venv/lib/python3.10/site-packages/libcst/_nodes/base.py", line 233, in visit
    visitor.on_leave(self)
  File "/home/steven/transformers/.venv/lib/python3.10/site-packages/libcst/_visitors.py", line 137, in on_leave
    leave_func(original_node)
  File "/home/steven/transformers/utils/modular_model_converter.py", line 1357, in leave_Module
    self.visited_modules[file] = ModelFileMapper.visit_and_merge_dependencies(
  File "/home/steven/transformers/utils/modular_model_converter.py", line 945, in visit_and_merge_dependencies
    wrapper.visit(mapper)
  File "/home/steven/transformers/.venv/lib/python3.10/site-packages/libcst/metadata/wrapper.py", line 204, in visit
    return self.module.visit(visitor)
  File "/home/steven/transformers/.venv/lib/python3.10/site-packages/libcst/_nodes/module.py", line 89, in visit
    result = super(Module, self).visit(visitor)
  File "/home/steven/transformers/.venv/lib/python3.10/site-packages/libcst/_nodes/base.py", line 227, in visit
    _CSTNodeSelfT, self._visit_and_replace_children(visitor)
  File "/home/steven/transformers/.venv/lib/python3.10/site-packages/libcst/_nodes/module.py", line 74, in _visit_and_replace_children
    body=visit_body_sequence(self, "body", self.body, visitor),
  File "/home/steven/transformers/.venv/lib/python3.10/site-packages/libcst/_nodes/internal.py", line 227, in visit_body_sequence
    return tuple(visit_body_iterable(parent, fieldname, children, visitor))
  File "/home/steven/transformers/.venv/lib/python3.10/site-packages/libcst/_nodes/internal.py", line 193, in visit_body_iterable
    new_child = child.visit(visitor)
  File "/home/steven/transformers/.venv/lib/python3.10/site-packages/libcst/_nodes/base.py", line 227, in visit
    _CSTNodeSelfT, self._visit_and_replace_children(visitor)
  File "/home/steven/transformers/.venv/lib/python3.10/site-packages/libcst/_nodes/statement.py", line 442, in _visit_and_replace_children
    body=visit_sequence(self, "body", self.body, visitor),
  File "/home/steven/transformers/.venv/lib/python3.10/site-packages/libcst/_nodes/internal.py", line 177, in visit_sequence
    return tuple(visit_iterable(parent, fieldname, children, visitor))
  File "/home/steven/transformers/.venv/lib/python3.10/site-packages/libcst/_nodes/internal.py", line 159, in visit_iterable
    new_child = child.visit(visitor)
  File "/home/steven/transformers/.venv/lib/python3.10/site-packages/libcst/_nodes/base.py", line 218, in visit
    should_visit_children = visitor.on_visit(self)
  File "/home/steven/transformers/.venv/lib/python3.10/site-packages/libcst/_visitors.py", line 123, in on_visit
    retval = visit_func(node)
  File "/home/steven/transformers/utils/modular_model_converter.py", line 657, in visit_ImportFrom
    import_module = self.python_module.code_for_node(node.module)
  File "/home/steven/transformers/.venv/lib/python3.10/site-packages/libcst/_nodes/module.py", line 136, in code_for_node
    node._codegen(state)
AttributeError: 'NoneType' object has no attribute '_codegen'
```

So I changed a bit the way I import the SuperGlueImageProcessor class by doing the following : 
```py 
from ..superglue import SuperGlueImageProcessor

class LightGlueImageProcessor(SuperGlueImageProcessor):
    pass
```
Doing so doesn't raise an error like above, but this is the resulting ``image_processing_lightglue.py`` file : 
```py
#                🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨
#           This file was automatically generated from src/transformers/models/lightglue/modular_lightglue.py.
#               Do NOT edit this file manually as any edits will be overwritten by the generation of
#             the file from the modular. If any change should be done, please apply the change to the
#                          modular_lightglue.py file directly. One of our CI enforces this.
#                🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨
# Copyright 2025 The HuggingFace Team. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from ..superglue import SuperGlueImageProcessor


class LightGlueImageProcessor(SuperGlueImageProcessor):
    pass
```

Is there something wrong in what I do ? Any ideas ? 🤔 

Anyway, thanks for coming back to my PR, I also add very little time recently to work on my PR's 😅

Cyrilvallez commented on src/transformers/models/lightglue/modular_lightglue.py:
```diff
@@ -0,0 +1,963 @@
+        return hidden_states
+
+
+class LightGlueAttentionBlock(nn.Module):
+    def __init__(self, config: LightGlueConfig, layer_idx: int):
+        super().__init__()
+        self.attention = LightGlueAttention(config, layer_idx)
+        self.mlp = LightGlueMLP(config)
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        position_embeddings: Optional[Tuple[torch.Tensor, torch.Tensor]] = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        encoder_hidden_states: Optional[torch.Tensor] = None,
+        encoder_attention_mask: Optional[torch.Tensor] = None,
+        output_hidden_states: bool = False,
+        output_attentions: bool = False,
+    ) -> Tuple[torch.Tensor, Optional[Tuple[torch.Tensor]], Optional[torch.Tensor]]:
+        all_hidden_states = () if output_hidden_states else None
+        attention_output, attention_probs = self.attention(
+            hidden_states,
+            position_embeddings=position_embeddings,
+            attention_mask=attention_mask,
+            encoder_hidden_states=encoder_hidden_states,
+            encoder_attention_mask=encoder_attention_mask,
+            output_attentions=output_attentions,
+        )
+        intermediate_states = torch.cat([hidden_states, attention_output], dim=-1)
+        output_states = self.mlp(intermediate_states)
+        hidden_states = hidden_states + output_states
+
+        if output_hidden_states:
+            all_hidden_states = all_hidden_states + (intermediate_states, output_states)
+
+        outputs = (hidden_states, all_hidden_states, attention_probs)
+        return outputs
```
If possible, we prefer to incorporate that directly to the TransformerLayer 🤗 If the weights were already converted and it's too annoying to convert again, it's fine though and we can let it as is - let me know!

Cyrilvallez commented on tests/models/lightglue/test_image_processing_lightglue.py:
```diff
@@ -0,0 +1,90 @@
+
+@require_torch
+@require_vision
+class LightGlueImageProcessingTest(SuperGlueImageProcessingTest, unittest.TestCase):
+    image_processing_class = LightGlueImageProcessor if is_vision_available() else None
```
You need to override the `setUp` method as well, to define the correct imega_processor_tester

Cyrilvallez commented on tests/models/lightglue/test_modeling_lightglue.py:
```diff
@@ -0,0 +1,584 @@
+
+    @slow
+    def test_model_from_pretrained(self):
+        from_pretrained_ids = ["stevenbucaille/lightglue_superpoint"]
```
If possible let's try to only use official checkpoints everywhere

sbucaille replied:
Ah yes, also got rid of other leftovers here and there and made sure it always links to the official checkpoints

Cyrilvallez reviewed (commented):
Alright! I asnwered your modular question above! Except for that part, and the tiny details I'm adding right now, this is super clean and ready to be added! 🤗 Great work 🤗
Let me know when you apply the last changes (and if modular works correctly by changing the import in superpoint), and then we'll be ready to merge

sbucaille committed:
refactor: update LightGlue model references

sbucaille committed:
fix: import error

sbucaille committed:
test: enhance LightGlue image processing tests with setup method

- Added a setup method in `LightGlueImageProcessingTest` to initialize `LightGlueImageProcessingTester`.
- Included a docstring for `LightGlueImageProcessingTester` to clarify its purpose.

sbucaille committed:
refactor: added LightGlue image processing implementation to modular file

sbucaille committed:
refactor: moved attention blocks into the transformer layer

sbucaille committed:
fix: added missing import

sbucaille committed:
fix: added missing import in __all__ variable

Cyrilvallez reviewed (approved):
Alright! Final nit comment! I also answered you above, let's add a small comment in the config to explain why we set to eager 👌
Once both changes are made, I'll merge! 🤗

sbucaille committed:
doc: added comment about enforcing eager attention because of SuperPoint

sbucaille committed:
refactor: added SuperPoint eager attention comment and moved functions to the closest they are used

sbucaille commented:
@Cyrilvallez thanks for the quick iteration, I've made the changes, added the comment agout SuperPoint eager attention and moved the functions down to where they belong, feel free to merge ! 🤗

Cyrilvallez reviewed (approved):
Alright, all good! Thanks a lot for bearing with us, and sorry again for the time it took between reviews! Great work! Merging now! 🤗🚀

qubvel commented:
Congrats on another model merged @sbucaille 🎉 Great wor, thanks for iterating on the model for a long time and not giving up 🤗 🚀
